kubernetes-sigs/aws-load-balancer-controller
# Add support for NLB instance mode

Fixes: #1744 

## Overview
The support for NLB instance mode is available from the k8s service controller, and any changes or improvements are tied to the upstream release schedule. In addition, there are also limitations on using CRDs in the in-tree controller.

This PR is for adding support for NLB instance mode in the AWS Load balancer controller. The controller already supports instance mode for ALB, and we will build up on  the `TargetGroupBinding` feature for NLB instance mode as well. 

Moving forward, the k8s AWS service controller will have support for NLB instance mode and CLB with basic features only. For more NLB features like the TargetGroupBinding, proxy v2 support, improved health checks, IP target mode we recommend running the AWS load balancer controller. Each service of type LoadBalancer on AWS k8s is reconciled either by the in-tree controller or the external controller based on the annotations provided the version requirements are met.

## k8s version requirement
The instance mode support for service type LoadBalancer requires k8s 1.20 (non-EKS) or later where we've modified the in-tree controller to ignore service with aws-load-balancer-type annotation of `external` or `nlb-ip`. For EKS, we've backported the changes to EKS 1.16 and later releases.

## How to configure load balancer target type?
If the service annotation `service.beta.kubernetes.io/aws-load-balancer-type` is external, this controller will look for an additional annotation `service.beta.kubernetes.io/aws-load-balancer-target-type` to determine the target type. The current supported values for the target types annotation are `nlb-instance` and `nlb-ip`.

### Instance target
Example service spec for instance target
```
kind: Service
apiVersion: v1
metadata:
  name: nlb-svc
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: "external"
    service.beta.kubernetes.io/aws-load-balancer-target-type: "nlb-instance"
spec:
  ports:
    - name: port
      port: 80
      targetPort: 80
      protocol: TCP
  type: LoadBalancer
  selector:
    app: instance-mode
```

### IP target
For consistency with instance target mode, IP mode will also be determined based on the following annotations - 
```
service.beta.kubernetes.io/aws-load-balancer-type: external
service.beta.kubernetes.io/aws-load-balancer-target-type: nlb-ip
```
For backwards compatibility, the `service.beta.kubernetes.io/aws-load-balancer-type` annotation value of `nlb-ip` will still be interpreted as IP mode NLB.

### Alternative considered
Modify the in-tree controller to support additional value for the `service.beta.kubernetes.io/aws-load-balancer-type` annotation, for example 
```
service.beta.kubernetes.io/aws-load-balancer-type: nlb-instance
```
Pros:
- Single annotation, and is similar to `nlb-ip`

Cons:
- We have to backport the change to all supported EKS versions, and will be dependent on the patch rollout. This results in different behavior across multiple platform versions and will be more confusing to the end users
- One more patch to the in-tree controller we need to maintain.

## Testing
- Create service of type Load Balancer with instance targets, External Traffic Policy set to Cluster
- Create service of type Load Balancer with instance targets, External Traffic Policy set to Local
- health check annotations get honored
- proxy protocol v2
- preserve client IP can be disabled for instance mode
- Security group rules get added as expected
- The labels specified in`service.beta.kubernetes.io/aws-load-balancer-target-node-labels` annotation works as expected
   - Without this annotation, the default filter gets applied
   - When service is created/edited with this annotaton, only matching nodes get registered
   - On edit, any nodes not matching the labels get de-registered from the target group
   - Node label edits are correctly reconciled
